[bug#12464] Fix resizing for `ArrayBuffer` and `PriorityQueue`

NthPortal opened the pull request:
Fix `Int` overflows when ensuring the size of an `ArrayBuffer`.

Always use `ensureSize` in `ArrayBuffer.from` to check
against the VM array size limit.

Make sizing behaviour of `ArrayBuffer` consistent.

Partially addresses scala/bug#12464 (still need to address `ArrayDeque`)

NthPortal committed:
[bug#12464] Fix resizing for `ArrayBuffer` and `PriorityQueue`

Fix `Int` overflows when ensuring the size of an `ArrayBuffer`.

Always use `ensureSize` in `ArrayBuffer.from` to check
against the VM array size limit.

Make sizing behaviour of `ArrayBuffer` consistent.

OndrejSpanel commented on src/library/scala/collection/mutable/ArrayBuffer.scala:
```diff
@@ -66,7 +66,13 @@ class ArrayBuffer[A] private (initialElements: Array[AnyRef], initialSize: Int)
   /** Ensure that the internal array has at least `n` cells. */
   protected def ensureSize(n: Int): Unit = {
-    array = ArrayBuffer.ensureSize(array, size0, n)
+    array = ArrayBuffer.ensureSize(array, size0, n.toLong)
```
`n.toLong` is not really necessary here, is it? In other places the conversion is done before adding, to make it does not overflow. Here is is just passed. Making the conversion explicit does no harm, though, and it makes it obvious what is going on.

NthPortal replied:
I mean, the method takes a `Long`, so I called `toLong`. I guess I forgot that the compiler would insert the conversion for me

OndrejSpanel commented on src/library/scala/collection/mutable/ArrayBuffer.scala:
```diff
@@ -66,7 +66,13 @@ class ArrayBuffer[A] private (initialElements: Array[AnyRef], initialSize: Int)
+  // TODO 3.T: should be `protected`, perhaps `protected[this]`
+  /** Ensure that the internal array has at least `n` additional cells more than `size0`. */
+  private[mutable] def ensureAdditionalSize(n: Int): Unit = {
+    array = ArrayBuffer.ensureSize(array, size0, size0.toLong + n)
```
Here .toLong is necessary. Perhaps a comment might be added to explain why? I am not sure if it is clean enough from the code.

NthPortal replied:
if you don't do `toLong` on one of them first, it adds the two `Int`s together, which has the same overflow problem. to me, at least, that's obvious? it's fairly common to call `to<number_type>` on one argument before doing maths, to ensure that the operation is done using the correct type (e.g. calling `toDouble` to ensure you get floating point division rather than integer division).

OndrejSpanel commented on src/library/scala/runtime/PStatics.scala:
```diff
@@ -0,0 +1,19 @@
+// things that should be in `Statics`, but can't be yet for bincompat reasons
+// TODO 3.T: move to `Statics`
+private[scala] object PStatics {
+  final val VM_MaxArraySize = 2147483645 // == `Int.MaxValue - 2`, hotspot limit
```
Is the hotspot limit documented somewhere? If it is, a link would come handy. I there any reason why `Int.MaxValue - 2` is only a comment? Why not writing `val VM_MaxArraySize = Int.MaxValue - 2` instead?

NthPortal replied:
I don't know where, if anywhere, the hotspot limit is documented. I used the exact value because I can never remember how much constant folding the compiler does, and I wanted to make sure it was a constant. I remember there being a limit to how much folding the compiler will do

OndrejSpanel replied:
I have found this: https://stackoverflow.com/questions/3038392/do-java-arrays-have-a-maximum-size . It seems VM dependent, a value of `Integer.MAX_VALUE - 8` is mentioned in one of the answers: https://stackoverflow.com/a/8381338/16673

NthPortal replied:
I could perhaps be convinced to change it to that. I'm more inclined to just pick the lowest value that anyone can find on a VM though. it also doesn't hurt to reduce it later if need be

SethTisue commented:
@NthPortal have the conversations above concluded? do you think (does @scala/collections think) this has now received sufficient review?

SethTisue commented:
Haha Dale just noticed that CI is failing. It's just MiMa, but of course that needs to be resolved.

NthPortal commented:
I don't have strong feelings about how to resolve the conversations. I'm inclined to leave the max array size as is, until shown otherwise, but have no feelings about the other. I've been quite under the weather for over a week now, and haven't really had the energy to work on anything. if someone wants to decide one way or another on the outstanding 2 conversations, I'm fine with that. I'm happy to fix MiMa and make changes based on the resolutions of the conversations, but I don't know how soon that would happen, so if someone else wants to, that's fine with me

SethTisue commented:
anyone interested in pursuing this further...?

NthPortal commented:
I'll try to get to it... this weekend maybe? working again, so I don't have a lot of capacity these days

NthPortal commented:
@SethTisue I've addressed all the comments (and added MiMa filters). unless anyone else wants to review, I think it's good to go once the build finishes

lrytz commented:
LGTM, but it would be good if another pair of eyes with brain behind could take a look.

SethTisue commented on src/library/scala/collection/mutable/ArrayBuffer.scala:
```diff
@@ -54,6 +53,7 @@ class ArrayBuffer[A] private (initialElements: Array[AnyRef], initialSize: Int)
 
   @transient private[this] var mutationCount: Int = 0
 
+  // needs to be private[collection] for parallel-collections
```
oops, this is out of sync with the code (`private` vs `protected`)

NthPortal replied:
@SethTisue that was actually intentional. `protected[x]` is the union of `protected` and `private[x]`. to fulfill the needs of the stdlib (`PriorityQueue` in particular), it must be `protected`. parallel-collections requires it to be `private[collection]`, but does not make use of it being `protected`.

SethTisue replied:
Okay — but the comment doesn't communicate that to the reader. As written, it _reads_ like a mistake.

NthPortal replied:
... stop making good points Seth

SethTisue commented:
Consider this approved for merge once my one tiny nitpick is fixed.
